src/pal/pal_apple.h

```diff
@@ -18,75 +28,158 @@ namespace snmalloc
     {
-      SNMALLOC_ASSERT(is_aligned_block<PALBSD::page_size>(p, size));
-#    ifdef USE_POSIX_COMMIT_CHECKS
+      SNMALLOC_ASSERT(is_aligned_block<page_size>(p, size));
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
       memset(p, 0x5a, size);
-#    endif
-      while (madvise(p, size, MADV_FREE_REUSABLE) == -1 && errno == EAGAIN)
-        ;
+#  endif
+
+      // `MADV_FREE_REUSABLE` can only be applied to writable pages,
+      // otherwise it's an error.
+      //
+      // `mach_vm_behavior_set` is observably slower in benchmarks.
+      madvise(p, size, MADV_FREE_REUSABLE);
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      // This must occur after `MADV_FREE_REUSABLE`.
+      //
+      // `mach_protect` is observably slower in benchmarks.
+      mprotect(p, size, PROT_NONE);
+#  endif
     }
 
     /**
-     * same remark as above but we need to mark the page as REUSE
-     * first
+     * Notify platform that we will be using these pages.
      */
     template<ZeroMem zero_mem>
     static void notify_using(void* p, size_t size) noexcept
     {
       SNMALLOC_ASSERT(
-        is_aligned_block<PALBSD::page_size>(p, size) || (zero_mem == NoZero));
-      while (madvise(p, size, MADV_FREE_REUSE) == -1 && errno == EAGAIN)
-        ;
+        is_aligned_block<page_size>(p, size) || (zero_mem == NoZero));
 
       if constexpr (zero_mem == YesZero)
-        zero<true>(p, size);
+      {
+        void* r = mmap(
+          p,
+          size,
+          PROT_READ | PROT_WRITE,
+          MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED,
+          anonymous_memory_fd,
+          0);
+
+        if (likely(r != MAP_FAILED))
+        {
+          return;
+        }
+      }
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      // Mark pages as writable for `madvise` below.
+      //
+      // `mach_vm_protect` is observably slower in benchmarks.
+      mprotect(p, size, PROT_READ | PROT_WRITE);
+#  endif
+
+      // `MADV_FREE_REUSE` can only be applied to writable pages,
+      // otherwise it's an error.
+      //
+      // `mach_vm_behavior_set` is observably slower in benchmarks.
+      madvise(p, size, MADV_FREE_REUSE);
+
+      if constexpr (zero_mem == YesZero)
+      {
+        ::bzero(p, size);
+      }
     }
+
+    template<bool committed>
+    static void* reserve_aligned(size_t size) noexcept
+    {
+      SNMALLOC_ASSERT(bits::is_pow2(size));
+      SNMALLOC_ASSERT(size >= minimum_alloc_size);
+
+      // mask has least-significant bits set
+      mach_vm_offset_t mask = size - 1;
+
+      int flags =
+        VM_FLAGS_ANYWHERE | VM_FLAGS_RANDOM_ADDR | default_mach_vm_map_flags;
+
+      // must be initialized to 0 or addr is interepreted as a lower-bound.
+      mach_vm_address_t addr = 0;
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      vm_prot_t prot = committed ? VM_PROT_READ | VM_PROT_WRITE : VM_PROT_NONE;
+#  else
+      vm_prot_t prot = VM_PROT_READ | VM_PROT_WRITE;
 #  endif
+
+      kern_return_t kr = mach_vm_map(
```

Comment:
does it still hold up with your recent changes ?

Reply:
Yes. This is necessary b/c there's no equivalent BSD/POSIX syscall. I was aiming for correctness and future-proofing.

ctest and mimalloc-bench report similar performance as master.

Reply:
Ok fair enough, I was just thinking about perf again.

Reply:
Please can you add a comment here explaining that this call is used because `mmap` doesn't provide the alignment.  @staceyson, I take it that `MAP_ALIGNED` is still not there on new XNU?